### PR TITLE
refactor(gtest, core): refactor weight dumping, dump additional constants

### PR DESCRIPTION
**Pull request by playX18**
~~Some constants are still there, need discussion whether to include them or no such as: `EPOCH_DURATION_IN_BLOCKS, EXISTENTIAL_DEPOSIT, GAS_ALLOWANCE,MAILBOX_THRESHOLD, MAX_RESERVATIONS, RESERVE_FOR, VALUE_PER_GAS`. We can include them into weight-diff util as well if that's desired. ~~

Me and @StackOverflowExcept1on refactored a lot of code to be more 1-to-1 with pallet gear schedule and core.
- Added ProcessCosts dump
- Added DbWeights type to gear pallet, moved `read` and `write` fields out of `Schedule` to `DbWeights`
- Moved codegened structures to gear-core
- Added test to verify correctness of ProcessCosts dump
- Moved instruction backward compatibility test to vara runtime integration tests

### CODE REVIEW

File: utils/weight-diff/src/main.rs [hunk 1]
```diff
@@ -437,6 +454,37 @@ fn main() {
...
+                }
+            });
+
+            let waitlist = dump.waitlist;
+            let dispatch_stash = dump.dispatch_stash;
+            let reservation = dump.reservation;
+            let read = dump.read;
+            let write = dump.write;
+
+            declarations.push(quote! {
```
Comment (StackOverflowExcept1on) [inline utils/weight-diff/src/main.rs:474]: it seems to be done automatically through struct visitor. you only need to add structure name 2 times (search in code for `"InstantiationWeights"`)

Comment (StackOverflowExcept1on) [reply utils/weight-diff/src/main.rs:474]: https://github.com/gear-tech/gear/blob/9482418e30c0b5fdd6644e3448271e1e3fe75407/utils/weight-diff/src/main.rs#L422-L430

File: utils/wasm-instrument/src/gas_metering/schedule.rs [hunk 2]
```diff
@@ -757,3 +757,23 @@ pub struct Weight {
     pub ref_time: u64,
     pub proof_size: u64,
 }
+
+pub struct ExtraWeights {
```
Comment (StackOverflowExcept1on) [inline utils/wasm-instrument/src/gas_metering/schedule.rs:761]: You need to modify `pallets/gear/src/schedule.rs` and add new weights there, and the `weight-diff` utility will dump the weights to `utils/wasm-instrument/src/gas_metering/schedule.rs`.

File: utils/weight-diff/src/main.rs (hunk 1 above)
Comment (playX18) [reply utils/weight-diff/src/main.rs]: I've added the fields to pallet gear Schedule so I assume modifications for weight-diff utility are not required anymore

File: utils/weight-diff/src/main.rs [hunk 3]
```diff
@@ -270,10 +275,17 @@ fn main() {
     match command {
         Commands::Dump { output_path, label } => {
             let writer = fs::File::create(output_path).unwrap();
+            let vara_schedule: Schedule<vara_runtime::Runtime> = Default::default();
+            let process_costs = vara_schedule.process_costs();
```
Comment (StackOverflowExcept1on) [inline utils/weight-diff/src/main.rs:279]: I wouldn't merge `Schedule` and `ProcessCosts` into one struct. We could dump `ProcessCosts` from the vara-runtime and add `serde` attributes to it. We might also have to add `serde` attributes to all `ProcessCosts` fields...

In this way we could implement the `process_costs` method, but in our lightweight scheduler (`utils/wasm-instrument/src/gas_metering/schedule.rs`).

File: pallets/gear/src/schedule.rs [hunk 4]
```diff
@@ -126,6 +126,9 @@ pub struct Schedule<T: Config> {
     /// The weights for memory interaction.
     pub memory_weights: MemoryWeights<T>,
 
+    /// Holding in storages rent costs.
```
Comment (StackOverflowExcept1on) [inline pallets/gear/src/schedule.rs:129]:
```suggestion
    /// The weights for renting.
```

File: pallets/gear/src/schedule.rs [hunk 5]
```diff
@@ -667,6 +670,19 @@ impl<T: Config> From<MemoryWeights<T>> for LazyPagesCosts {
     }
 }
 
+impl From<InstantiationWeights> for InstantiationCosts {
+    fn from(val: InstantiationWeights) -> Self {
+        Self {
+            code_section_per_byte: val.code_section_per_byte.ref_time().into(),
```
Comment (StackOverflowExcept1on) [inline pallets/gear/src/schedule.rs:676]: use the same field order as in InstantiationWeights

File: pallets/gear/src/schedule.rs [hunk 6]
```diff
@@ -690,6 +706,41 @@ pub struct InstantiationWeights {
     pub type_section_per_byte: Weight,
 }
 
+/// Describes the weight for renting.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Clone, Encode, Decode, PartialEq, Eq, WeightDebug, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct RentWeights<T: Config> {
+    pub dispatch_stash: Weight,
```
Comment (StackOverflowExcept1on) [inline pallets/gear/src/schedule.rs:714]: same for the order of the fields. use the same order as in RentCosts

File: gsdk/src/metadata/generated.rs [hunk 7]
```diff
@@ -2551,19 +2551,28 @@ pub mod runtime_types {
                     pub upload_page_data: runtime_types::sp_weights::weight_v2::Weight,
                     pub mem_grow: runtime_types::sp_weights::weight_v2::Weight,
                     pub mem_grow_per_page: runtime_types::sp_weights::weight_v2::Weight,
+                    pub read: runtime_types::sp_weights::weight_v2::Weight,
+                    pub write: runtime_types::sp_weights::weight_v2::Weight,
+                    pub read_per_byte: runtime_types::sp_weights::weight_v2::Weight,
+                    pub write_per_byte: runtime_types::sp_weights::weight_v2::Weight,
```
Comment (StackOverflowExcept1on) [inline gsdk/src/metadata/generated.rs:2557]: this is db/storage weights, not memory

File: utils/weight-diff/src/main.rs [hunk 8]
```diff
@@ -264,6 +269,97 @@ impl<'ast> Visit<'ast> for StructuresVisitor {
...
+    "InstructionCosts",
+    "SyscallCosts",
+    "MemoryCosts",
+    "RentCosts",
+    "InstantiationCosts",
+    "LazyPagesCosts",
+];
+
+impl<'ast> Visit<'ast> for ImplementationVisitor {
+    fn visit_item_impl(&mut self, node: &'ast syn::ItemImpl) {
```
Comment (StackOverflowExcept1on) [inline utils/weight-diff/src/main.rs:287]: mb split this patching small methods?

File: utils/weight-diff/src/main.rs [hunk 9]
```diff
@@ -360,11 +456,22 @@ fn main() {
...
+                .collect::<Vec<_>>();
+
             let mut declarations = vec![quote! {
                 //! This is auto-generated module that contains cost schedule from
                 //! `pallets/gear/src/schedule.rs`.
                 //!
                 //! See `./scripts/weight-dump.sh` if you want to update it.
+
+                use crate::costs::*;
             }];
```
Comment (StackOverflowExcept1on) [inline utils/weight-diff/src/main.rs:475]: to add \n
```suggestion
            let mut declarations = vec![quote! {
                //! This is auto-generated module that contains cost schedule from
                //! `pallets/gear/src/schedule.rs`.
                //!
                //! See `./scripts/weight-dump.sh` if you want to update it.
            }];

            declarations.push(quote! {
                use crate::costs::*;
            });
```

**Review by StackOverflowExcept1on — APPROVED** [review]

**Review by StackOverflowExcept1on — APPROVED** [review]

**Review by breathx — APPROVED** [review]

**Pull request merged**